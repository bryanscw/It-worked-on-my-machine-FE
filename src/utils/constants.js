export const API_URL = 'http://localhost:3007';

export const USER_TYPES = {
    STUDENT: 'STUDENT',
    TEACHER: 'TEACHER',
    ADMIN: 'ADMIN',
};

export const CREATE = 'CREATE';
export const UPDATE = 'UPDATE';
export const DELETE = 'DELETE';
export const EMPTY = 'EMPTY';