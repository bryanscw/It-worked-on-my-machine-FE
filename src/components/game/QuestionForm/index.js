import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { EMPTY } from '../../../utils/constants';

/**
 * This component displays the Question interface in the Popup
 */

export class QuestionForm extends Component {
    emptyState = {
        answer: null,
    }
    state = this.emptyState

    componentDidMount() {
        const {
            initialState
        } = this.props;
        
        if (initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);

            else
                this.setState(initialState);
        }
    }

    componentDidUpdate(prevProps) {
        const {
            initialState
        } = this.props;

        if (initialState && initialState !== prevProps.initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);
                
            else
                this.setState(initialState);
        }
    }

    onSubmit = e => {
        e.preventDefault();

        const {
            answer,
        } = this.state;

        this.props.onSubmit(answer);

        this.setState({
            answer: null
        });
    }

    onChangeAnswer = answer => {
        this.setState({
            answer: answer
        });
    }

    render() {
        const {
            options
        } = this.props;

        const {
            answer,
        } = this.state;

        return (
            <form className="form" onSubmit={this.onSubmit}>
                <p><strong>Options</strong></p>
                {
                    Object.keys(options).map(key => {
                        let option = options[key];
                        
                        return (
                            <div className="form-check" key={key}>
                                <input
                                    className="form-check-input"
                                    id={`option-${key}`}
                                    type="radio"
                                    name="radio"
                                    onChange={e => this.onChangeAnswer(e.target.checked ? key : null)}
                                    checked={answer === key}
                                    />
                                <label className="form-check-label" htmlFor={`option-${key}`}>{option}</label>
                            </div>
                        )
                    })
                }
                <div className="form-group mt-4">
                    <button className="btn btn-primary" type="submit">Submit</button>
                </div>
            </form>
        )
    }
}

QuestionForm.propTypes = {
    /** The initial state of the form. If it's 'EMPTY', then the form will be empty. If it's an object, the form will display the form with initial state's content. */
    initialState: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    /** The function to call when the form is submitted */
    onSubmit: PropTypes.func.isRequired,
}

export default QuestionForm
