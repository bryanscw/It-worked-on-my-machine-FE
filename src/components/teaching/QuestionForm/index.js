import React, { Component } from 'react';
import { EMPTY } from '../../../utils/constants';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faTrash } from '@fortawesome/free-solid-svg-icons';

export class QuestionForm extends Component {
    emptyState = {
        questionText: '',
        options: [],
        answer: null,
    };
    state = this.emptyState;

    componentDidMount() {
        const {
            initialState
        } = this.props;

        if (initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);

            else
                this.setState(initialState);
        }
    }

    componentDidUpdate(prevProps) {
        const {
            initialState
        } = this.props;

        if (initialState && initialState !== prevProps.initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);
                
            else
                this.setState(initialState);
        }
    }

    onSubmit = e => {
        e.preventDefault();

        const {
            questionText,
            options,
            answer,
        } = this.state;

        this.props.onSubmit({
            questionText,
            options,
            answer,
        });

        this.setState(this.emptyState);
    }

    onChange = e => {
        this.setState({
            [e.target.name]: e.target.value
        });
    }

    onChangeOption = (selectedIndex, value) => {
        this.setState({
            options: this.state.options.map((option, i) => i === selectedIndex ? value : option),
        });
    }

    onChangeAnswer = answer => {
        this.setState({
            answer: answer
        });
    }

    onCreateOption = () => {
        this.setState({
            options: [...this.state.options, ""]
        });
    }

    onDeleteOption = selectedIndex => {
        let answer = this.state.answer;

        if (selectedIndex < answer) {
            answer--;
        } else if (selectedIndex === answer) {
            answer = null;
        }

        this.setState({
            options: this.state.options.filter((option, i) => i !== selectedIndex),
            answer: answer,
        });
    }

    render() {
        const {
            questionText,
            options,
            answer,
        } = this.state;

        return (
            <form className="form" onSubmit={this.onSubmit}>
                <div className="form-group">
                    <label>Question Text</label>
                    <input
                        className="form-control"
                        type="text"
                        name="questionText"
                        onChange={this.onChange}
                        value={questionText}
                        />
                </div>
                <p><strong>Options (Select the Correct Option)</strong></p>
                {
                    options.map((option, i) => 
                        <div className="form-group d-flex" key={i}>
                            <div className="form-check mr-1">
                                <input
                                    className="form-check-input"
                                    type="radio"
                                    name="radio"
                                    onChange={e => this.onChangeAnswer(e.target.checked ? i : null)}
                                    checked={answer === parseInt(i)}
                                    />
                            </div>
                            <input
                                className="form-control w-100 mr-2"
                                type="text"
                                name="option"
                                onChange={e => this.onChangeOption(i, e.target.value)}
                                value={option}
                                />
                            <button className="btn btn-danger btn-sm" type="button" onClick={() => this.onDeleteOption(i)}>
                                <FontAwesomeIcon icon={faTrash}/>
                            </button>
                        </div>
                    )
                }
                <button className="btn btn-primary" type="button" onClick={() => this.onCreateOption()}>
                    <FontAwesomeIcon icon={faPlus} className="mr-2" />Add Option
                </button>
                <div className="form-group mt-4">
                    <button className="btn btn-primary" type="submit">Submit</button>
                </div>
            </form>
        )
    }
}

export default QuestionForm
