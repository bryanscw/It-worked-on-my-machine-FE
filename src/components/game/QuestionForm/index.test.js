import React from 'react'
import { render, cleanup } from '@testing-library/react'
import QuestionForm from './index.js';
import { EMPTY } from '../../../utils/constants.js';

afterEach(cleanup);

it('should take a snapshot', () => {
    const { asFragment } = render(
        <QuestionForm
            options={{
                "0": 1,
                "1": 2,
            }}
            initialState={EMPTY}
            onSubmit={() => {}}
            />
    );
    expect(asFragment()).toMatchSnapshot();
});
