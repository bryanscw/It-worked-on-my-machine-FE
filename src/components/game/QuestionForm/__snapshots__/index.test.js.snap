// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <p>
      <strong>
        Options
      </strong>
    </p>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="option-0"
        name="radio"
        type="radio"
      />
      <label
        class="form-check-label"
        for="option-0"
      >
        1
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="option-1"
        name="radio"
        type="radio"
      />
      <label
        class="form-check-label"
        for="option-1"
      >
        2
      </label>
    </div>
    <div
      class="form-group mt-4"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
